### DISCUSSION TITLE
v5 roadmap

### DISCUSSION THREAD

**Discussion by mistic100** [mistic100/Photo-Sphere-Viewer#593] [General]
# Publish as ES6 modules

Three.js, since a few versions already, exclusively publishes ES6 code to NPM, even the big `three.min.js` uses `const` and `class` and so on.  
I think it's time for PSV to do the same.  
People using the lastest version of three.js but targeting old browsers must already have babel in their build-chain so it shouln't break anything. And people only using direct `script` tags (like the demos) are already not working on old browsers.

Thinks to consider :

- check the used syntax, things like `.?` and `??` might not have landed on all browsers already ; will probably need to keep babel to do the job
- what to publish ? three.js publish a UMD file (`three.js`) and an ES6 module (`three.module.js` but still one big file)
- change the architecture of the package : currently files are in the `dist` folder, it makes import of plugins ugly
- remove deprecated stuff (a lot of code in `utils` was written excludively for old browsers compatibility)
- drop uEvent and use native events instead (how to handle [changer events](https://github.com/mistic100/uEvent#change) ?)

# Publish in multiple NPM packages ?

- @psv/core
- @psv/adapter-***
- @psv/plugin-***

# Change position parameters names

`longitude` and `latitude` have always been source of confusion, being often understood as GPS coordinates. They should be renamed :

- longitude => angleH or positionX
- latitude => angleV or positionY
- x => panoramaX
- y => panoramaY

This will reflect in all APIs using angles as well as markers and virtual tour configuration (here "position" should be renamed in "gps" or something)

# Migrate to TypeScript

Because it's better! And maintaining d.ts files is cumbersome.

**Comment by thangdjw** [comment]
i think we need moving effects

**Reply by mistic100** [reply]
please elaborate

**Reply by thangdjw** [reply]
I think you need to add smooth travel, transition between points. you can refer to 3dvista

**Reply by mistic100** [reply]
This is an Open Source project, if you want a new feature you must create an issue with a maximum of details. "Look at this other product" is not a proper request.

BTW there is already a keypoints plugin https://photo-sphere-viewer.js.org/plugins/plugin-autorotate-keypoints.html

**Comment by beginor** [comment]
Yes, I think it's time to move to es6 modules!

Since webgl is a feature of modern browser, so there is no need to consider the legacy old browser .

**Reply by mistic100** [reply]
That's not really true... WebGL works perfectly on IE 11 and this library too (requires some transpiling for three.js and a bunch of polyfills).
